pkg/csi/service/vanilla/controller.go

```diff
@@ -2274,6 +2267,291 @@ func (c *controller) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRe
+				if err != nil {
+					return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to check if cns snapshot is supported on VC %s due to error: %v", vcHost, err)
+				}
+				if !isCnsSnapshotSupported {
+					return nil, "", logger.LogNewErrorCodef(log, codes.Unimplemented,
+						"VC %s version does not support snapshot operations", vcHost)
+				}
+				// fetch all snapshot records for each VC
+				snapshotQueryFilter = cnstypes.CnsSnapshotQueryFilter{
+					Cursor: &cnstypes.CnsCursor{
+						Offset: 0,
+						Limit:  utils.DefaultQuerySnapshotLimit,
+					},
+				}
+				for {
+					queryResult, err := volManager.QuerySnapshots(ctx, snapshotQueryFilter)
+					if err != nil {
+						log.Errorf("querySnapshots failed for vCenter %s with snapfilter %v. Err=%+v",
+							vcHost, snapshotQueryFilter, err)
+						return nil, "", err
+					}
+					perVCSnapQueryEntries = append(perVCSnapQueryEntries, queryResult.Entries...)
+					// If cursor offset < total snapshot records found, then there are more snapshots
+					// to be queried, so continue the loop to fetch remaining records
+					if queryResult.Cursor.Offset >= queryResult.Cursor.TotalRecords {
+						break
+					}
+					// Assign cursor for next iteration based on previous query result
+					snapshotQueryFilter.Cursor = &queryResult.Cursor
+				}
+				//populate list of volume-ids to retrieve the volume size.
+				var volumeIds []cnstypes.CnsVolumeId
+				for _, queryResult := range perVCSnapQueryEntries {
+					if queryResult.Error != nil {
+						return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+							"faults are not expected when invoking QuerySnapshots without volume-id and snapshot-id, fault: %+v",
+							queryResult.Error.Fault)
+					}
+					volumeIds = append(volumeIds, queryResult.Snapshot.VolumeId)
+				}
+				// TODO: Retrieve Snapshot size directly from CnsQuerySnapshot once supported.
+				volumeDetails, err := utils.QueryVolumeDetailsUtil(ctx, volManager, volumeIds)
+				if err != nil {
+					log.Errorf("failed to retrieve volume details for volume-ids: %v for vCenter %s, err: %+v",
+						volumeIds, vcHost, err)
+					return nil, "", err
+				}
+				snapQueryEntries = append(snapQueryEntries, perVCSnapQueryEntries...)
+				cnsVolumeDetailsMap = append(cnsVolumeDetailsMap, volumeDetails)
+			}
+			CNSSnapshotsForListSnapshots = snapQueryEntries
+			CNSVolumeDetailsMap = cnsVolumeDetailsMap
+		} else {
+			// Check for snapshot support
+			isCnsSnapshotSupported, err := vCenterManager.IsCnsSnapshotSupported(ctx,
+				c.manager.VcenterConfig.Host)
+			if err != nil {
+				return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to check if cns snapshot is supported on VC due to error: %v", err)
+			}
+			if !isCnsSnapshotSupported {
+				return nil, "", logger.LogNewErrorCode(log, codes.Unimplemented,
+					"VC version does not support snapshot operations")
+			}
```

Comment:
Create a wrapper for this, the code is called repeatedly in several places.

Reply:
done